zebra-chain/src/sapling/tests/prop.rs

```diff
@@ -2,33 +2,51 @@ use proptest::prelude::*;
 
 use crate::{
     block,
-    sapling::{self, PerSpendAnchor},
+    sapling::{self, PerSpendAnchor, SharedAnchor},
     serialization::{ZcashDeserializeInto, ZcashSerialize},
     transaction::{LockTime, Transaction},
 };
 
 use futures::future::Either;
 
 proptest! {
-    // TODO: generalise this test for `ShieldedData<SharedAnchor>` (#1829)
+
+    // Serialize and deserialize `PerSpendAnchor` and `SharedAnchor` shielded data
+    //  by including them in V4 and V5 transactions respectivly.
     #[test]
-    fn shielded_data_roundtrip(shielded in any::<sapling::ShieldedData<PerSpendAnchor>>()) {
+    fn shielded_data_roundtrip(
+        shielded_v4 in any::<sapling::ShieldedData<PerSpendAnchor>>(),
+        shielded_v5 in any::<sapling::ShieldedData<SharedAnchor>>()
+    ) {
         zebra_test::init();
 
         // shielded data doesn't serialize by itself, so we have to stick it in
         // a transaction
+
+        // stick `PerSependAnchor` shielded data into a v4 transaction
         let tx = Transaction::V4 {
             inputs: Vec::new(),
             outputs: Vec::new(),
             lock_time: LockTime::min_lock_time(),
             expiry_height: block::Height(0),
             joinsplit_data: None,
-            sapling_shielded_data: Some(shielded),
+            sapling_shielded_data: Some(shielded_v4),
         };
-
         let data = tx.zcash_serialize_to_vec().expect("tx should serialize");
         let tx_parsed = data.zcash_deserialize_into().expect("randomized tx should deserialize");
+        prop_assert_eq![tx, tx_parsed];
 
+        // stick `SharedAnchor` shielded data into a v5 transaction
+        let tx = Transaction::V5 {
+            lock_time: LockTime::min_lock_time(),
+            expiry_height: block::Height(0),
+            inputs: Vec::new(),
+            outputs: Vec::new(),
+            sapling_shielded_data: Some(shielded_v5),
+            rest: Vec::new(),
+        };
+        let data = tx.zcash_serialize_to_vec().expect("tx should serialize");
+        let tx_parsed = data.zcash_deserialize_into().expect("randomized tx should deserialize");
```

Comment:
This is failing. Unsure if i am doing something wrong or if there is a bug. It fails here: https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/sapling/keys.rs#L915

It seems that some points generated are outside of the curve or something like that, i did some debugging and found that many times the `Choice` generated by the strategy is true while the application crash at some points where it is false. 

```
[zebra-chain/src/sapling/keys.rs:912] possible_point = CtOption {
    value: AffinePoint {
        u: 0x0000000000000000000000000000000000000000000000000000000000000000,
        v: 0x0000000000000000000000000000000000000000000000000000000000000001,
    },
    is_some: Choice(
        1,
    ),
}
[zebra-chain/src/sapling/keys.rs:912] possible_point = CtOption {
    value: AffinePoint {
        u: 0x0000000000000000000000000000000000000000000000000000000000000000,
        v: 0x0000000000000000000000000000000000000000000000000000000000000001,
    },
    is_some: Choice(
        1,
    ),
}
[zebra-chain/src/sapling/keys.rs:912] possible_point = CtOption {
    value: AffinePoint {
        u: 0x0000000000000000000000000000000000000000000000000000000000000000,
        v: 0x0000000000000000000000000000000000000000000000000000000000000001,
    },
    is_some: Choice(
        1,
    ),
}
[zebra-chain/src/sapling/keys.rs:912] possible_point = CtOption {
    value: AffinePoint {
        u: 0x0000000000000000000000000000000000000000000000000000000000000000,
        v: 0x0000000000000000000000000000000000000000000000000000000000000001,
    },
    is_some: Choice(
        1,
    ),
}
[zebra-chain/src/sapling/keys.rs:912] possible_point = CtOption {
    value: AffinePoint {
        u: 0x0000000000000000000000000000000000000000000000000000000000000000,
        v: 0x0000000000000000000000000000000000000000000000000000000000000001,
    },
    is_some: Choice(
        1,
    ),
}
[zebra-chain/src/sapling/keys.rs:912] possible_point = CtOption {
    value: AffinePoint {
        u: 0x0000000000000000000000000000000000000000000000000000000000000000,
        v: 0x0000000000000000000000000000000000000000000000000000000000000001,
    },
    is_some: Choice(
        1,
    ),
}
[zebra-chain/src/sapling/keys.rs:912] possible_point = CtOption {
    value: AffinePoint {
        u: 0x542a945450872686feaef24e8d138916177496774c24ad0dc28d0ea40d687faf,
        v: 0x1c00000000000000000000000000000000000000000000000000000000000000,
    },
    is_some: Choice(
        1,
    ),
}
[zebra-chain/src/sapling/keys.rs:912] possible_point = CtOption {
    value: AffinePoint {
        u: 0x0000000000000000000000000000000000000000000000000000000000000000,
        v: 0x0abf61a0e20ae30a743eebd6380765c92fdf02fc000000000000000000000000,
    },
    is_some: Choice(
        0,
    ),
}
The application panicked (crashed).
Message:  randomized tx should deserialize: Parse("Invalid jubjub::AffinePoint value")
```

Any help is appreciated.

Reply:
This is a serialisation problem - the serialise and deserialise impls for Spend don't match. And the overall transaction v5 serialisation and deserialisation don't match the spec.

See the spec and comment: https://github.com/ZcashFoundation/zebra/pull/1996#pullrequestreview-633062418